kernel/work_q.c

```diff
@@ -132,12 +132,12 @@ int k_delayed_work_submit_to_queue(struct k_work_q *work_q,
 
 int k_delayed_work_cancel(struct k_delayed_work *work)
 {
-	if (!work->work_q) {
-		return -EINVAL;
-	}
-
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	int ret = work_cancel(work);
+	int ret = z_abort_timeout(&work->timeout);
```

Comment:
That is not the equivalent to what work_cancel does:

```
		/* Remove from the queue if already submitted */
		if (!k_queue_remove(&work->work_q->queue, &work->work)) {
			return -EINVAL;
		}
```

I see you are suggesting to remove this behavior but I do recall people asking for such a change so they can cancel a work no matter it is already been queued to execution or it is still pending on timeout.

Reply:
It's a change in behavior, but not a change in documented API.

The original work queue architecture never supported removing an item from a queue once it had been submitted.  Only delayed work items could be canceled, and the documentation has always (since 2016) said this could only be done before the timer elapsed:

```
 * @brief Cancel a delayed work item.
 *
 * This routine cancels the submission of delayed work item @a work.
 * A delayed work item can only be canceled while its countdown is still
 * underway.
 *
```

The ability to remove an item from a work queue was only added for delayed work items, and the implementation was not robust: there's no way to distinguish multiple states so the return value could not correctly indicate whether the cancellation was successful.

Reply:
I wrote that since by the time we could not remove items from the queue, it was a fifoe, but this has since change we now can as the code I posted before was doing.

The part of not being robust really puzzle me, should we actually attempt to make it robust instead of just remove it completely, or do you believe the feature of cancelling pending itens is not necessary, because I think it is and we do have users for that.

Reply:
Btw, you could have come with a test that proves your point about it not being robust enough, because so far I haven't seem an issue with the robustness of this code and there is a lot code using k_delayed_work_cancel not just bluetooth and no issue has been created, well other than yours.

Reply:
>  the feature of cancelling pending

This feature makes workqueue usable. The fact that application will have to track how many times submission was executed will make things more complicated for everybody and lead to bugs.
The cancel if done from the right context was safe. We cancel works from workqueue thread context but I believe this could be done from any thread as long as workqueue thread is cooperative.